affectedmoduledetector/src/main/kotlin/com/dropbox/affectedmoduledetector/ToStringLogger.kt

```diff
@@ -20,52 +20,95 @@
 
 package com.dropbox.affectedmoduledetector
 
-import org.gradle.BuildAdapter
-import org.gradle.BuildResult
-import org.gradle.api.invocation.Gradle
+import org.gradle.api.Project
+import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logger
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.build.event.BuildEventListenerRegistryInternal
 import org.gradle.internal.logging.slf4j.OutputEventListenerBackedLogger
 import org.gradle.internal.logging.slf4j.OutputEventListenerBackedLoggerContext
+import org.gradle.internal.operations.*
 import org.gradle.internal.time.Clock
+import org.gradle.invocation.DefaultGradle
+import java.io.File
 
 /**
  * Gradle logger that logs to a string.
  */
-class ToStringLogger(
-    private val stringBuilder: StringBuilder = StringBuilder()
+internal open class ToStringLogger(
+    private val stringBuilder: Provider<ToStringLoggerBuildService>?
 ) : OutputEventListenerBackedLogger(
         "amd",
-        OutputEventListenerBackedLoggerContext(
-                Clock {
-                    System.currentTimeMillis()
-                }
-        ).also {
+        OutputEventListenerBackedLoggerContext {
+            System.currentTimeMillis()
+        }.also {
             it.level = LogLevel.DEBUG
-            it.setOutputEventListener {
-                stringBuilder.appendln(it.toString())
+            it.setOutputEventListener { outputEvent ->
+                stringBuilder?.get()?.parameters?.getStringBuilderProperty()?.get()?.appendLine(outputEvent.toString())
             }
         },
         Clock {
             System.currentTimeMillis()
         }
 ) {
+
     /**
      * Returns the current log.
      */
-    fun buildString() = stringBuilder.toString()
+    fun buildString() = stringBuilder?.get()?.parameters?.getStringBuilderProperty()?.get()?.toString()
 
+    @Suppress("UnstableApiUsage") // BuildService is not yet stable
     companion object {
+        internal abstract class ToStringLoggerBuildService : BuildService<ToStringLoggerBuildService.ToStringLoggerBuildServiceParameters>, BuildOperationListener, AutoCloseable {
```

Comment:
I didn't undestand why abstract class in companion object space? Abstract class anyway cant be with different instances.

Reply:
It is abstract because `BuildService` classes are supposed to be abstract https://docs.gradle.org/current/userguide/build_services.html